File: scripts/import-client-data.ts
// Import client data using Neon HTTP driver (port 443 - never blocked)
// Has retry logic + 400ms throttle to avoid Neon rate limiting

import { neon } from '@neondatabase/serverless';
import fs from 'fs';
import { v5 as uuidv5 } from 'uuid';

const NAMESPACE = '6ba7b812-9dad-11d1-80b4-00c04fd430c8';
const JSON_FILE = 'temp_scraper/q84sale_listings.json';

const sql = neon(process.env.DATABASE_URL!);

function hasArabic(text?: string | null) {
  return text ? /[\u0600-\u06FF]/.test(text) : false;
}
function mapCondition(cond?: string | null) {
  if (!cond) return 'good';
  if (cond.toUpperCase() === 'NEW') return 'new';
  return 'good';
}
function parseDate(dateStr?: string | null): string {
  if (!dateStr) return new Date().toISOString();
  const d = new Date(dateStr);
  return isNaN(d.getTime()) ? new Date().toISOString() : d.toISOString();
}
function getSellerUuid(name: string) {
  const safe = (name || '').toLowerCase().replace(/[^a-z0-9]/g, '') || 'unknownseller';
  return uuidv5(`seller_${safe}`, NAMESPACE);
}
function getProductUuid(adId: string) { return uuidv5(`ad_${adId}`, NAMESPACE); }
function getImageUuid(productUuid: string, imgUrl: string) {
  return uuidv5(`${productUuid}_${imgUrl.slice(-80)}`, NAMESPACE);
}
const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

async function upsertListing(listing: any) {
  const adId = listing.ad_id?.toString();
  if (!adId) return 'skip';

  const sellerName = listing.seller_name || 'Unknown Seller';
  const sellerUuid = getSellerUuid(sellerName);
  const email = `${sellerUuid.slice(0, 8)}@q84sale.local`;
  const phone = `+96500${sellerUuid.slice(0, 6)}`;
  const productUuid = getProductUuid(adId);
  const rawTitle = listing.title || '';
  const rawDesc = listing.description || '';
  const isAr = hasArabic(rawTitle);
  const price = listing.price ?? 0;
  const currency = listing.currency || 'KWD';
  const condition = mapCondition(listing.condition);
  const createdAt = parseDate(listing.posted_date_estimated);

  await sql`
    INSERT INTO users (id, email, password_hash, name, phone, country_code, is_verified, is_active, updated_at)
    VALUES (${sellerUuid}, ${email}, 'scraper_dump', ${sellerName}, ${phone}, 'KW', true, true, NOW())
    ON CONFLICT (email) DO UPDATE SET updated_at = NOW()
  `;
  await sql`
    INSERT INTO sellers (user_id, business_name, phone_public, is_verified, is_profile_complete, updated_at)
    VALUES (${sellerUuid}, ${sellerName}, ${phone}, true, true, NOW())
    ON CONFLICT (user_id) DO UPDATE SET updated_at = NOW()
  `;
  await sql`
    INSERT INTO products (id, seller_id, title, title_ar, description, description_ar, price, currency, condition, country_code, status, created_at, updated_at)
    VALUES (
      ${productUuid}, ${sellerUuid},
      ${rawTitle}, ${isAr ? rawTitle : null},
      ${rawDesc || null}, ${isAr ? rawDesc || null : null},
      ${price}, ${currency}, ${condition}, 'KW', 'active',
      ${createdAt}::timestamptz, NOW()
    )
    ON CONFLICT (id) DO UPDATE SET price = EXCLUDED.price, updated_at = NOW()
  `;
  const images: string[] = listing.images || [];
  for (let j = 0; j < images.length; j++) {
    const imgUuid = getImageUuid(productUuid, images[j]);
    await sql`
      INSERT INTO product_images (id, product_id, url, s3_key, is_primary, sort_order)
      VALUES (${imgUuid}, ${productUuid}, ${images[j]}, 'PENDING', ${j === 0}, ${j})
      ON CONFLICT (id) DO NOTHING
    `;
  }
  return 'ok';
}

async function main() {
  console.log(`\nLoading ${JSON_FILE}...`);
  const items: any[] = JSON.parse(fs.readFileSync(JSON_FILE, 'utf-8'));
  console.log(`Loaded ${items.length} listings.\n`);

  console.log('Testing DB connection...');
  const test = await sql`SELECT COUNT(*) as count FROM products`;
  console.log(`✅ Connected! Current products in DB: ${test[0].count}\n`);

  let ok = 0, skipped = 0, errors = 0;

  for (let i = 0; i < items.length; i++) {
    const adId = items[i].ad_id?.toString();
    if (!adId) { skipped++; continue; }

    let success = false;
    for (let attempt = 1; attempt <= 3; attempt++) {
      try {
        const result = await upsertListing(items[i]);
        if (result === 'skip') { skipped++; } else { ok++; }
        success = true;
        break;
      } catch (err: any) {
        if (attempt < 3) {
          await sleep(2000 * attempt); // 2s then 4s backoff
        } else {
          errors++;
          console.error(`  [${i + 1}] ❌ ad ${adId}: ${err.message?.slice(0, 80)}`);
        }
      }
    }

    if ((i + 1) % 50 === 0 || i === items.length - 1) {
      console.log(`  [${i + 1}/${items.length}] ✅ Imported: ${ok} | Errors: ${errors}`);
    }

    await sleep(400); // 400ms throttle between items
  }

  console.log('\n========================================');
  console.log('IMPORT COMPLETE');
  console.log('========================================');
  console.log(`✅ Successfully imported: ${ok}`);
  console.log(`⏭️  Skipped:              ${skipped}`);
  console.log(`❌ Errors:               ${errors}`);
  const final = await sql`SELECT COUNT(*) as count FROM products`;
  console.log(`📦 Total products in DB: ${final[0].count}`);
  console.log('========================================\n');
}

main().catch((e) => { console.error(e); process.exit(1); });

File: scripts/import_client_data.py
import json
import uuid
import os
import psycopg2
from psycopg2.extras import execute_values
from datetime import datetime, timedelta
import re
from dotenv import load_dotenv

load_dotenv()

DB_URI = os.getenv("DATABASE_URL")
if not DB_URI:
    raise ValueError("DATABASE_URL not set. Please create a .env file with DATABASE_URL=postgresql://...")
JSON_FILE = "temp_scraper/q84sale_listings.json"

def has_arabic(text):
    if not text:
        return False
    return bool(re.search(r'[\u0600-\u06FF]', text))

def map_condition(cond_str):
    if not cond_str:
        return 'good'
    cond_str = cond_str.upper()
    if cond_str == 'NEW':
        return 'new'
    if cond_str == 'USED':
        return 'good'
    return 'good'

def parse_date(date_str):
    if not date_str:
        return datetime.now()
    try:
        return datetime.fromisoformat(date_str)
    except:
        return datetime.now()

def main():
    print(f"Loading data from {JSON_FILE}...")
    try:
        with open(JSON_FILE, 'r', encoding='utf-8') as f:
            data = json.load(f)
    except Exception as e:
        print(f"Error loading {JSON_FILE}: {e}")
        return

    print(f"Loaded {len(data)} items. Connecting to DB...")
    try:
        conn = psycopg2.connect(DB_URI)
        cursor = conn.cursor()
    except Exception as e:
        print(f"Error connecting to database: {e}")
        return

    # Metrics
    sellers_upserted = 0
    products_upserted = 0
    images_upserted = 0
    errors = []

    # Simple text to id converter for sellers
    def get_seller_id(name):
        safe_name = re.sub(r'[^a-zA-Z0-9]', '', str(name).lower())
        if not safe_name:
            safe_name = "unknownseller"
        return str(uuid.uuid5(uuid.NAMESPACE_OID, f"seller_{safe_name}"))

    for listing in data:
        ad_id = str(listing.get("ad_id", ""))
        if not ad_id:
            continue
            
        seller_name = listing.get("seller_name", "Unknown Seller")
        seller_uuid = get_seller_id(seller_name)
        
        safe_email = f"{seller_uuid[:8]}@q84sale.local"
        dummy_phone = f"+9650000{seller_uuid[:4]}"
        
        # 1. Upsert User 
        try:
            cursor.execute("""
                INSERT INTO users (id, email, password_hash, name, phone, country_code, is_verified, is_active, updated_at)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, NOW())
                ON CONFLICT (email) DO UPDATE 
                SET updated_at = NOW()
            """, (seller_uuid, safe_email, "scraper_dump", seller_name, dummy_phone, "KW", True, True))
        except Exception as e:
            errors.append(f"Error user {seller_name}: {e}")
            conn.rollback()
            continue

        # 2. Upsert Seller
        try:
            cursor.execute("""
                INSERT INTO sellers (user_id, business_name, phone_public, is_verified, is_profile_complete, updated_at)
                VALUES (%s, %s, %s, %s, %s, NOW())
                ON CONFLICT (user_id) DO UPDATE 
                SET updated_at = NOW()
            """, (seller_uuid, seller_name, dummy_phone, True, True))
            if cursor.rowcount > 0:
                sellers_upserted += 1
        except Exception as e:
            errors.append(f"Error seller {seller_name}: {e}")
            conn.rollback()
            continue
            
        # 3. Upsert Product
        product_uuid = str(uuid.uuid5(uuid.NAMESPACE_OID, f"ad_{ad_id}"))
        
        raw_title = listing.get("title", "")
        raw_desc = listing.get("description", "")
        
        is_ar = has_arabic(raw_title) or has_arabic(raw_desc)
        title_ar = raw_title if is_ar else None
        title_en = raw_title if not is_ar else raw_title # fallback
        desc_ar = raw_desc if is_ar else None
        desc_en = raw_desc if not is_ar else raw_desc # fallback
        
        price = listing.get("price") or 0
        currency = listing.get("currency") or "KWD"
        condition = map_condition(listing.get("condition"))
        
        post_date_estimated = listing.get("posted_date_estimated")
        created_at = parse_date(post_date_estimated)
        
        try:
            cursor.execute("""
                INSERT INTO products (id, seller_id, title, title_ar, description, description_ar, price, currency, condition, country_code, status, created_at, updated_at)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, 'active', %s, NOW())
                ON CONFLICT (id) DO UPDATE 
                SET price = EXCLUDED.price,
                    status = EXCLUDED.status,
                    updated_at = NOW()
            """, (product_uuid, seller_uuid, title_en, title_ar, desc_en, desc_ar, price, currency, condition, "KW", created_at))
            
            if cursor.rowcount > 0:
                products_upserted += 1
        except Exception as e:
            errors.append(f"Error product {ad_id}: {e}")
            conn.rollback()
            continue
            
        # 4. Upsert Images
        images = listing.get("images", [])
        for i, img_url in enumerate(images):
            # Using MD5 of URL to ensure unique IDs per image
            img_uuid = str(uuid.uuid5(uuid.NAMESPACE_OID, f"{product_uuid}_{img_url}"))
            try:
                cursor.execute("""
                    INSERT INTO product_images (id, product_id, url, s3_key, is_primary, sort_order)
                    VALUES (%s, %s, %s, %s, %s, %s)
                    ON CONFLICT (id) DO NOTHING
                """, (img_uuid, product_uuid, img_url, "PENDING", i == 0, i))
                
                if cursor.rowcount > 0:
                    images_upserted += 1
            except Exception as e:
                errors.append(f"Error image for {ad_id}: {e}")
                conn.rollback()

        conn.commit()

    conn.close()

    print("="*40)
    print("IMPORT SUMMARY")
    print("="*40)
    print(f"Sellers upserted:  {sellers_upserted}")
    print(f"Products upserted: {products_upserted}")
    print(f"Images upserted:   {images_upserted}")
    if errors:
        print(f"Errors:            {len(errors)}")
        for err in errors[:5]:
            print(f"  - {err}")
    print("="*40)

if __name__ == "__main__":
    main()
